src/workerd/api/node/zlib-util.h

```diff
@@ -250,38 +282,111 @@ class ZlibUtil final: public jsg::Object {
+    }
+
+    void initializeStream(jsg::BufferSource _write_result, jsg::Function<void()> writeCallback);
+
+    // Allocation functions provided to zlib itself. We store the real size of
+    // the allocated memory chunk just before the "payload" memory we return
+    // to zlib.
+    static void* AllocForZlib(void* data, uInt items, uInt size);
+    static void* AllocForBrotli(void* data, size_t size);
+    static void FreeForZlib(void* data, void* pointer);
+
+  private:
+    CompressionContext context_;
+    bool initialized = false;
+    bool writing = false;
+    bool pending_close = false;
+    bool closed = false;
+
+    // Equivalent to `write_js_callback` in Node.js
+    jsg::Optional<jsg::Function<void()>> writeCallback;
+    jsg::Optional<jsg::BufferSource> writeResult;
+    jsg::Optional<CompressionStreamErrorHandler> errorHandler;
+
+    // Used to store allocations in Brotli* operations.
+    kj::HashMap<uint8_t*, kj::Array<uint8_t>> allocations;
+  };
+
+  class ZlibStream final: public CompressionStream<ZlibContext> {
+  public:
+    explicit ZlibStream(ZlibMode mode): CompressionStream(mode) {}
+    KJ_DISALLOW_COPY_AND_MOVE(ZlibStream);
+    static jsg::Ref<ZlibStream> constructor(ZlibModeValue mode);
+
+    // Instance methods
+    void initialize(int windowBits,
+        int level,
+        int memLevel,
+        int strategy,
+        jsg::BufferSource writeState,
+        jsg::Function<void()> writeCallback,
+        jsg::Optional<kj::Array<kj::byte>> dictionary);
+    void params(jsg::Lock& js, int level, int strategy);
+
     JSG_RESOURCE_TYPE(ZlibStream) {
+      JSG_INHERIT(CompressionStream<ZlibContext>);
+
+      JSG_METHOD(initialize);
+      JSG_METHOD(params);
+    }
+  };
+
+  template <typename CompressionContext>
+  class BrotliCompressionStream: public CompressionStream<CompressionContext> {
+  public:
+    explicit BrotliCompressionStream(ZlibMode _mode)
+        : CompressionStream<CompressionContext>(_mode) {}
+    KJ_DISALLOW_COPY_AND_MOVE(BrotliCompressionStream);
+    static jsg::Ref<BrotliCompressionStream> constructor(ZlibModeValue mode);
+
+    bool initialize(jsg::Lock& js,
+        jsg::BufferSource params,
+        jsg::BufferSource writeResult,
+        jsg::Function<void()> writeCallback);
+
+    void params() {
+      // Currently a no-op, and not accessed from JS land.
+      // At some point Brotli may support changing parameters on the fly,
+      // in which case we can implement this and a JS equivalent similar to
+      // the zlib Params() function.
+    }
+
+    JSG_RESOURCE_TYPE(BrotliCompressionStream) {
+      JSG_INHERIT(CompressionStream<CompressionContext>);
+
       JSG_METHOD(initialize);
-      JSG_METHOD(close);
-      JSG_METHOD(write);
-      JSG_METHOD(writeSync);
       JSG_METHOD(params);
-      JSG_METHOD(setErrorHandler);
-      JSG_METHOD(reset);
+    }
+
+    CompressionContext* context() {
+      return this->CompressionStream<CompressionContext>::context();
```

Comment:
Not sure what this method does because it should be inherited.

Reply:
it returns the context of the inherited class. This won't work because the usage of excessive templates makes the compilers work extremely hard when used with `this->context.myFunction()`